File: solutions/scala/123-sliding-window-maximum.scala
/**
 * Problem 123: Sliding Window Maximum (LeetCode 239)
 * Difficulty: Hard
 * Language: Scala
 */
import scala.collection.immutable.Queue
def maxSlidingWindow(nums: Array[Int], k: Int): Array[Int] = {
  val (_, res) = nums.indices.foldLeft((Queue.empty[Int], Array.empty[Int])) {
    case ((dq, res), i) =>
      val dq1 = if (dq.nonEmpty && dq.head <= i - k) dq.tail else dq
      val dq2 = dq1.reverse.dropWhile(j => nums(j) <= nums(i)).reverse.enqueue(i)
      val r = if (i >= k - 1) res :+ nums(dq2.head) else res
      (dq2, r)
  }
  res
}

File: solutions/haskell/123-sliding-window-maximum.hs
{-
  Problem 123: Sliding Window Maximum (LeetCode 239)
  Difficulty: Hard
  Language: Haskell
-}
import Data.Sequence (Seq, (|>), ViewL(..), ViewR(..), viewl, viewr, empty)

maxSlidingWindow :: [Int] -> Int -> [Int]
maxSlidingWindow nums k =
  let arr = zip [0..] nums
      step (dq, res) (i, v) =
        let dq1 = case viewl dq of
              (j :< rest) | j <= i - k -> rest
              _ -> dq
            popR d = case viewr d of
              (rest :> j) | nums !! j <= v -> popR rest
              _ -> d
            dq2 = popR dq1 |> i
            res' = if i >= k - 1 then res ++ [nums !! front]
                   else res
            front = case viewl dq2 of (j :< _) -> j
        in (dq2, res')
  in snd $ foldl step (empty, []) arr
